### PR TITLE
Fix bug #5599

### PR DESCRIPTION
**Pull request by hoangviet1985**
Solves polynominal equations more correctly.

### CODE REVIEW

File: modules/core/src/mathfuncs.cpp [hunk 1]
```diff
@@ -2511,12 +2509,46 @@ double cv::solvePoly( InputArray _coeffs0, OutputArray _roots0, int maxIters )
...
+            if( num_same_root > 1)
+            {
+                for( j = 0; j < num_same_root / 2; j++)
+                {
+                    double old_num_re = num.re;
+                    double old_num_im = num.im;
+
+                    num.re = old_num_re*old_num_re + old_num_im*old_num_im;
+                    Mat num_re_arr = (Mat_<double>(1,1) << num.re);
+                    cv::sqrt(num_re_arr, num_re_arr);
```
Comment (vpisarev) [inline modules/core/src/mathfuncs.cpp]: this is terribly inefficient - allocate 1x1 matrix and call cv::sqrt(). just use std::sqrt() instead

File: modules/core/src/mathfuncs.cpp [hunk 2]
```diff
@@ -2511,12 +2509,46 @@ double cv::solvePoly( InputArray _coeffs0, OutputArray _roots0, int maxIters )
...
+                    num.re = num_re_arr.at<double>(0, 0);
+                    num.re += old_num_re;
+                    num.im = num.re - old_num_re;
+                    num.re /= 2;
+                    num_re_arr.at<double>(0, 0) = num.re;
+                    cv::sqrt(num_re_arr, num_re_arr);
```
Comment (vpisarev) [inline modules/core/src/mathfuncs.cpp]: same

File: modules/core/src/mathfuncs.cpp [hunk 3]
```diff
@@ -2511,12 +2509,46 @@ double cv::solvePoly( InputArray _coeffs0, OutputArray _roots0, int maxIters )
...
+                    num.re = num_re_arr.at<double>(0, 0);
+
+                    Mat num_im_arr = (Mat_<double>(1,1) << num.im);
+                    num.im /= 2;
+                    num_im_arr.at<double>(0, 0) = num.re;
+                    cv::sqrt(num_im_arr, num_im_arr);
```
Comment (vpisarev) [inline modules/core/src/mathfuncs.cpp]: same

**Comment by vpisarev** [comment]
@hoangviet1985, thank you for the bug fix! besides a few small comments above, may I ask you to add a small regression sample that proves that the bug is fixed. And this would also help to check this automatically in the future

**Comment by hoangviet1985** [comment]
I added a file to test 2 cases, which are equations x^4 - x^2 = 0 and x^2 - 2x + 1 = 0. When solving these equations using current solvePoly function, returned roots are "nan".

**Comment by hoangviet1985** [comment]
@vpisarev , thank you for your ideas.

**Comment by hoangviet1985** [comment]
@vpisarev , there were unsuccessful builds on Linux. It said that because "no space left on device". I don't know this happened. Could you please explain? Thank you.

**Comment by alalek** [comment]
`/tmp` folder has been full. Trash is removed and build is restarted.

**Comment by hoangviet1985** [comment]
@alalek , thank you!

**Comment by vpisarev** [comment]
@hoangviet1985, the sample indeed demonstrates the problem, but I would prefer this sample to be converted into a test, because we run samples few times a year at best, some of the samples I have not run for a few years already, but the tests are executed on a daily basis. Could you please make a test and put it to modules/core/test/...?

**Comment by hoangviet1985** [comment]
@vpisarev , you want me to merge my test to the current test of solvePoly, or to create a separate test file?

**Comment by hoangviet1985** [comment]
@vpisarev , testing function of solvePoly method does not look complete to me. Would you like me to edit it in this pull request?

**Comment by alalek** [comment]
Just put into `modules/core/test/test_math.cpp` something like this:
```
template<typename T>
static void checkRoot(Mat& r, T re, T im)
{
    for (size_t i = 0; i < r.total(); i++)
    {
        Vec<T, 2> v = *(Vec<T, 2>*)r.ptr(i);
        if (fabs(re - v[0]) < 1e-6 && fabs(im - v[1]) < 1e-6)
        {
            v[0] = std::numeric_limits<T>::quiet_NaN();
            v[1] = std::numeric_limits<T>::quiet_NaN();
            return;
        }
    }
    GTEST_NONFATAL_FAILURE_("Can't find root") << "(" << re << ", " << im << ")";
}
TEST(Core_SolvePoly, regression_5599)
{
    // x^4 - x^2 = 0, roots: 1, -1, 0, 0
    cv::Mat coefs = (cv::Mat_<float>(1,5) << 0, 0, -1, 0, 1 );
    {
        cv::Mat r;
        double prec;
        prec = cv::solvePoly(coefs, r);
        EXPECT_LE(prec, 1e-6);
        EXPECT_EQ(4, (int)r.total());
        //std::cout << "Preciseness = " << prec << std::endl;
        //std::cout << "roots:\n" << r << "\n" << std::endl;
        ASSERT_EQ(CV_32FC2, r.type());
        checkRoot<float>(r, 1, 0);
        checkRoot<float>(r, -1, 0);
        checkRoot<float>(r, 0, 0);
        checkRoot<float>(r, 0, 0);
    }
    ... other cases ...
}

```

After adding please squash commits into one (whitespace fixups history is not needed)

**Comment by hoangviet1985** [comment]
@alalek , thank you. You just made my day.

**Comment by hoangviet1985** [comment]
I don't know why Mac build failed. Can someone help me? Thank you.

**Comment by vpisarev** [comment]
:+1:

**Pull request merged**